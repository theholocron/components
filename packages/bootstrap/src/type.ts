import * as React from "react";

export interface WithChildren {
	children?: React.ReactNode;
}
