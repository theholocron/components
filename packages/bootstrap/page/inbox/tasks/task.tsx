import { ActionIcon, Checkbox, Group, TextInput } from "@mantine/core";
import { IconStar, IconTrash } from "@tabler/icons-react";

export interface TaskProps {
	task: {
		id: string;
		title: string;
		state: "TASK_INBOX" | "TASK_PINNED" | "TASK_ARCHIVED";
	};
	onArchiveTask: (archive: "ARCHIVE_TASK", id: string) => void;
	onTogglePinTask: (
		state: "TASK_INBOX" | "TASK_PINNED" | "TASK_ARCHIVED",
		id: string
	) => void;
	onEditTitle: (title: string, id: string) => void;
	onDeleteTask: (id: string) => void;
}

export function Task(props: TaskProps) {
	const {
		task: { id, title, state },
		onArchiveTask,
		onTogglePinTask,
		onEditTitle,
		onDeleteTask,
	} = props;

	return (
		<Group
			align="center"
			justify="space-between"
			wrap="nowrap"
			gap="xl"
			role="listitem"
			aria-label={`task-${id}`}
		>
			<Checkbox
				type="checkbox"
				aria-label={`archiveTask-${id}`}
				onChange={() => onArchiveTask("ARCHIVE_TASK", id)}
				name="checked"
				id={`archiveTask-${id}`}
				checked={state === "TASK_ARCHIVED"}
				pl={10}
			/>

			<TextInput
				variant="unstyled"
				type="text"
				value={title}
				name="title"
				placeholder="Input title"
				style={{
					textOverflow: "ellipsis",
					width: "100%",
				}}
				onChange={(e) => onEditTitle(e.target.value, id)}
			/>
			<ActionIcon
				variant="transparent"
				onClick={() => onDeleteTask(id)}
				aria-label="delete"
			>
				<IconTrash
					style={{ width: "70%", height: "70%" }}
					stroke={1.5}
				/>
			</ActionIcon>
			{state !== "TASK_ARCHIVED" && (
				<ActionIcon
					variant="transparent"
					onClick={() => onTogglePinTask(state, id)}
					id={`pinTask-${id}`}
					aria-label={state === "TASK_PINNED" ? "unpin" : "pin"}
					key={`pinTask-${id}`}
					mr={10}
				>
					<IconStar
						style={{ width: "70%", height: "70%" }}
						stroke={1.5}
					/>
				</ActionIcon>
			)}
		</Group>
	);
}
