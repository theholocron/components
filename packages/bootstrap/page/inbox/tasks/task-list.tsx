import { Container, Group, Paper, Skeleton, Text, Title } from "@mantine/core";
import * as React from "react";
import { Task, type TaskProps } from "./task";
import { ITask } from "./use-tasks";

export interface TaskListProps {
	loading?: boolean;
	tasks: ITask[];
	onTogglePinTask: TaskProps["onTogglePinTask"];
	onArchiveTask: TaskProps["onArchiveTask"];
	onEditTitle: TaskProps["onEditTitle"];
	onDeleteTask: TaskProps["onDeleteTask"];
}

export function TaskList(props: TaskListProps) {
	const {
		loading = false,
		tasks,
		onTogglePinTask,
		onArchiveTask,
		onEditTitle,
		onDeleteTask,
	} = props;

	const events = {
		onTogglePinTask,
		onArchiveTask,
		onEditTitle,
		onDeleteTask,
	};

	const LoadingRow = (
		<Group
			align="center"
			justify="space-between"
			wrap="nowrap"
			gap="xl"
			role="listitem"
			mt={10}
			mb={10}
		>
			<Skeleton ml={10} width={20} height={20} />
			<Skeleton width="100%" height={32} />
			<Skeleton width={28} height={28} />
			<Skeleton mr={10} width={28} height={28} />
		</Group>
	);

	if (loading) {
		return (
			<React.Fragment>
				{LoadingRow}
				{LoadingRow}
				{LoadingRow}
				{LoadingRow}
				{LoadingRow}
				{LoadingRow}
			</React.Fragment>
		);
	}

	if (tasks.length === 0) {
		return (
			<Container size={420} my={40}>
				<Paper withBorder shadow="md" p={30} mt={30} radius="md">
					<Title ta="center">You have no tasks</Title>
					<Text c="dimmed" size="sm" ta="center" mt={5}>
						Sit back and relax
					</Text>
				</Paper>
			</Container>
		);
	}

	const tasksInOrder = [
		...tasks.filter((t) => t.state === "TASK_PINNED"),
		...tasks.filter((t) => t.state !== "TASK_PINNED"),
	];

	return (
		<div
			data-testid="success"
			key={"success"}
			role="list"
			aria-label="tasks"
		>
			{tasksInOrder.map((task) => (
				<Task key={task.id} task={task} {...events} />
			))}
		</div>
	);
}
