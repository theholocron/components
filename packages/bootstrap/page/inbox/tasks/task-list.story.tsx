import type { Meta, StoryObj } from "@storybook/react";
import * as TaskStories from "./task.story";
import { TaskList } from "./task-list";

const meta: Meta<typeof TaskList> = {
	component: TaskList,
	title: "TaskList",
	argTypes: {
		...TaskStories.default.argTypes,
	},
};
export default meta;
type Story = StoryObj<typeof TaskList>;

export const Default: Story = {
	args: {
		tasks: [
			{ id: "1", state: "TASK_INBOX", title: "Build a date picker" },
			{ id: "2", state: "TASK_INBOX", title: "QA dropdown" },
			{
				id: "3",
				state: "TASK_INBOX",
				title: "Write a schema for account avatar component",
			},
			{ id: "4", state: "TASK_INBOX", title: "Export logo" },
			{
				id: "5",
				state: "TASK_INBOX",
				title: "Fix bug in input error state",
			},
			{
				id: "6",
				state: "TASK_INBOX",
				title: "Draft monthly blog to customers",
			},
		],
	},
};

export const WithPinnedTasks: Story = {
	args: {
		tasks: [
			{
				id: "6",
				title: "Draft monthly blog to customers",
				state: "TASK_PINNED",
			},
			...(Default.args || { tasks: [] }).tasks.slice(0, 5),
		],
	},
};

export const Loading: Story = {
	args: {
		tasks: [],
		loading: true,
	},
};

export const Empty: Story = {
	args: {
		...Loading.args,
		loading: false,
	},
};
