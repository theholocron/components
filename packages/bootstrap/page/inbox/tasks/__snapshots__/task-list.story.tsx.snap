// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList Default smoke-test 1`] = `
<div class="list-items"
     data-testid="success"
     role="list"
     aria-label="tasks"
>
  <div class="list-item TASK_INBOX"
       role="listitem"
       aria-label="task-1"
  >
    <label for="checked"
           aria-label="archiveTask-1"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-1"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-1"
      >
      </span>
    </label>
    <label for="title"
           aria-label="Build a date picker"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="Build a date picker"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-1"
            aria-label="pin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
  <div class="list-item TASK_INBOX"
       role="listitem"
       aria-label="task-2"
  >
    <label for="checked"
           aria-label="archiveTask-2"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-2"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-2"
      >
      </span>
    </label>
    <label for="title"
           aria-label="QA dropdown"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="QA dropdown"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-2"
            aria-label="pin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
  <div class="list-item TASK_INBOX"
       role="listitem"
       aria-label="task-3"
  >
    <label for="checked"
           aria-label="archiveTask-3"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-3"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-3"
      >
      </span>
    </label>
    <label for="title"
           aria-label="Write a schema for account avatar component"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="Write a schema for account avatar component"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-3"
            aria-label="pin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
  <div class="list-item TASK_INBOX"
       role="listitem"
       aria-label="task-4"
  >
    <label for="checked"
           aria-label="archiveTask-4"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-4"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-4"
      >
      </span>
    </label>
    <label for="title"
           aria-label="Export logo"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="Export logo"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-4"
            aria-label="pin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
  <div class="list-item TASK_INBOX"
       role="listitem"
       aria-label="task-5"
  >
    <label for="checked"
           aria-label="archiveTask-5"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-5"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-5"
      >
      </span>
    </label>
    <label for="title"
           aria-label="Fix bug in input error state"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="Fix bug in input error state"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-5"
            aria-label="pin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
  <div class="list-item TASK_INBOX"
       role="listitem"
       aria-label="task-6"
  >
    <label for="checked"
           aria-label="archiveTask-6"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-6"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-6"
      >
      </span>
    </label>
    <label for="title"
           aria-label="Draft monthly blog to customers"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="Draft monthly blog to customers"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-6"
            aria-label="pin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
</div>
`;

exports[`TaskList Empty smoke-test 1`] = `
<div class="list-items"
     data-testid="empty"
>
  <div class="wrapper-message">
    <span class="icon-check">
    </span>
    <p class="title-message">
      You have no tasks
    </p>
    <p class="subtitle-message">
      Sit back and relax
    </p>
  </div>
</div>
`;

exports[`TaskList Loading smoke-test 1`] = `
<div class="list-items"
     data-testid="loading"
>
  <div class="loading-item">
    <span class="glow-checkbox">
    </span>
    <span class="glow-text">
      <span>
        Loading
      </span>
      <span>
        cool
      </span>
      <span>
        state
      </span>
    </span>
  </div>
  <div class="loading-item">
    <span class="glow-checkbox">
    </span>
    <span class="glow-text">
      <span>
        Loading
      </span>
      <span>
        cool
      </span>
      <span>
        state
      </span>
    </span>
  </div>
  <div class="loading-item">
    <span class="glow-checkbox">
    </span>
    <span class="glow-text">
      <span>
        Loading
      </span>
      <span>
        cool
      </span>
      <span>
        state
      </span>
    </span>
  </div>
  <div class="loading-item">
    <span class="glow-checkbox">
    </span>
    <span class="glow-text">
      <span>
        Loading
      </span>
      <span>
        cool
      </span>
      <span>
        state
      </span>
    </span>
  </div>
  <div class="loading-item">
    <span class="glow-checkbox">
    </span>
    <span class="glow-text">
      <span>
        Loading
      </span>
      <span>
        cool
      </span>
      <span>
        state
      </span>
    </span>
  </div>
  <div class="loading-item">
    <span class="glow-checkbox">
    </span>
    <span class="glow-text">
      <span>
        Loading
      </span>
      <span>
        cool
      </span>
      <span>
        state
      </span>
    </span>
  </div>
</div>
`;

exports[`TaskList WithPinnedTasks smoke-test 1`] = `
<div class="list-items"
     data-testid="success"
     role="list"
     aria-label="tasks"
>
  <div class="list-item TASK_PINNED"
       role="listitem"
       aria-label="task-6"
  >
    <label for="checked"
           aria-label="archiveTask-6"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-6"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-6"
      >
      </span>
    </label>
    <label for="title"
           aria-label="Draft monthly blog to customers"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="Draft monthly blog to customers"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-6"
            aria-label="unpin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
  <div class="list-item TASK_INBOX"
       role="listitem"
       aria-label="task-1"
  >
    <label for="checked"
           aria-label="archiveTask-1"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-1"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-1"
      >
      </span>
    </label>
    <label for="title"
           aria-label="Build a date picker"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="Build a date picker"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-1"
            aria-label="pin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
  <div class="list-item TASK_INBOX"
       role="listitem"
       aria-label="task-2"
  >
    <label for="checked"
           aria-label="archiveTask-2"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-2"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-2"
      >
      </span>
    </label>
    <label for="title"
           aria-label="QA dropdown"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="QA dropdown"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-2"
            aria-label="pin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
  <div class="list-item TASK_INBOX"
       role="listitem"
       aria-label="task-3"
  >
    <label for="checked"
           aria-label="archiveTask-3"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-3"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-3"
      >
      </span>
    </label>
    <label for="title"
           aria-label="Write a schema for account avatar component"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="Write a schema for account avatar component"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-3"
            aria-label="pin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
  <div class="list-item TASK_INBOX"
       role="listitem"
       aria-label="task-4"
  >
    <label for="checked"
           aria-label="archiveTask-4"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-4"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-4"
      >
      </span>
    </label>
    <label for="title"
           aria-label="Export logo"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="Export logo"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-4"
            aria-label="pin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
  <div class="list-item TASK_INBOX"
       role="listitem"
       aria-label="task-5"
  >
    <label for="checked"
           aria-label="archiveTask-5"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-5"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-5"
      >
      </span>
    </label>
    <label for="title"
           aria-label="Fix bug in input error state"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="Fix bug in input error state"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-5"
            aria-label="pin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
</div>
`;
